src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_adapter.py

```diff
@@ -0,0 +1,750 @@
+# Copyright 2022 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import numpy as np
+import torch
+from PIL import Image as PIL_Image
+from transformers import CLIPFeatureExtractor, CLIPTextModel, CLIPTokenizer
+
+from ...models import Adapter, AutoencoderKL, MultiAdapter, UNet2DConditionModel
+from ...schedulers import KarrasDiffusionSchedulers
+from ...utils import (
+    PIL_INTERPOLATION,
+    is_accelerate_available,
+    is_accelerate_version,
+    logging,
+    randn_tensor,
+    replace_example_docstring,
+)
+from ..pipeline_utils import DiffusionPipeline
+from . import StableDiffusionPipelineOutput
+from .safety_checker import StableDiffusionSafetyChecker
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> from PIL import Image
+        >>> from diffusers.utils import load_image
+
+        >>> image = load_image("https://huggingface.co/RzZ/sd-v1-4-adapter-color/resolve/main/color_ref.png")
+
+        >>> color_palette = image.resize((8, 8))
+        >>> color_palette = color_palette.resize((512, 512), resample=Image.Resampling.NEAREST)
+
+        >>> import torch
+        >>> from diffusers import StableDiffusionAdapterPipeline, Adapter
+
+        >>> adapter = Adapter.from_pretrained("RzZ/sd-v1-4-adapter-color")
+        >>> pipe = StableDiffusionAdapterPipeline.from_pretrained(
+        ...     "CompVis/stable-diffusion-v1-4",
+        ...     adapter=adapter,
+        ...     torch_dtype=torch.float16,
+        ... )
+
+        >>> pipe.to("cuda")
+
+        >>> out_image = pipe(
+        ...     "At night, glowing cubes in front of the beach",
+        ...     image=color_palette,
+        ...     generator=generator,
+        ... ).images[0]
+        ```
+"""
+
+
+def preprocess(image, target_height, target_width):
+    if isinstance(image, torch.Tensor):
+        return image
+    elif isinstance(image, PIL_Image.Image):
+        image = [image]
+
+    multi_control = isinstance(image[0], list) or isinstance(image[0], tuple)
+    multi_control |= isinstance(image[0], torch.Tensor) and image[0].ndim == 4
+
+    if multi_control:
+        images = [preprocess(subset, target_height, target_width) for subset in image]
+        b, c, h, w = images[0].shape
+        image = [img.reshape([b * c, h, w]) for img in images]
```

Comment:
general diffusers style is to avoid recursion if possible. Could we re-write this so there is a single call to preprocess? 

Also, ideally we don't have to check if there are multiple controls are being used. At least in the controlnet pre-processing code, we don't need to check

Reply:
I updated it so the pipeline now decides how to preprocess the images by looking at the instance type of `self.adapter`, so there is no more recursive inside `preprocess`, just like controlnet pipeline.